File: backend/routers/lobby.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from game.state_manager import create_session, get_session, add_player, update_state
from ai.llm_client import generate_country_resources
from routers.ws import broadcast

router = APIRouter(prefix="/game", tags=["lobby"])


class CreateGameRequest(BaseModel):
    player_name: str


class JoinGameRequest(BaseModel):
    player_name: str


class StartGameRequest(BaseModel):
    player_id: str


@router.post("/create")
async def create_game(body: CreateGameRequest):
    """Create a new game session. Returns game_id and the creator's player_id."""
    game_id, player_id = create_session(body.player_name)
    return {"game_id": game_id, "player_id": player_id}


@router.post("/join/{game_id}")
async def join_game(game_id: str, body: JoinGameRequest):
    """Join an existing pending session. Returns the new player's player_id."""
    player_id = add_player(game_id, body.player_name)
    if not player_id:
        raise HTTPException(
            status_code=400,
            detail="Cannot join: game not found, already started, or lobby is full.",
        )
    return {"player_id": player_id}


@router.post("/start/{game_id}")
async def start_game(game_id: str, body: StartGameRequest):
    print("start_game", game_id, body)
    """
    Start the game. Only the creator can call this.
    Triggers AI resource generation for all countries, then broadcasts game_started.

    NOTE: Requires at least 1 player (creator). For real multiplayer require >= 2.
    """
    state = get_session(game_id)
    if not state:
        raise HTTPException(status_code=404, detail="Game not found.")
    if state.creator_id != body.player_id:
        raise HTTPException(status_code=403, detail="Only the creator can start the game.")
    if state.status != "pending":
        raise HTTPException(status_code=400, detail="Game has already started.")

    # Generate resources for all countries via AI
    resources = await generate_country_resources({name: {} for name in state.countries})
    for country_name, res in resources.items():
        if country_name in state.countries:
            state.countries[country_name].favorite_resource = res["favorite_resource"]
            state.countries[country_name].hated_resource = res["hated_resource"]

    state.status = "active"
    update_state(game_id, state)

    await broadcast(game_id, {"type": "game_started", "state": state.model_dump(mode="json")})
    return {"status": "started"}

File: backend/src/routes/lobby.ts
import { Router } from "express";
import { Request, Response } from "express";
import {
  createSession,
  getSession,
  addPlayer,
  updateState,
} from "../game/stateManager";
import { generateCountryResources } from "../ai/llmClient";

export const lobbyRoutes = Router();

type CreateGameRequest = {
  playerName: string;
};

type JoinGameRequest = {
  playerName: string;
};

type StartGameRequest = {
  playerId: string;
};

export const createGame = async (req: Request, res: Response) => {
  const body: CreateGameRequest = req.body;
  const { gameId, playerId } = createSession(body.playerName);
  res.json({ gameId: gameId, playerId: playerId });
};

export const joinGame = async (req: Request, res: Response) => {
  const gameId = req.params.gameId;
  const body: JoinGameRequest = req.body;

  const playerId = addPlayer(gameId, body.playerName);
  if (!playerId) {
    res.status(400).json({
      error: "Cannot join: game not found, already started, or lobby is full.",
    });
    return;
  }

  res.json({ playerId: playerId });
};

export const startGame = async (req: Request, res: Response) => {
  const gameId = req.params.gameId;
  const body: StartGameRequest = req.body;

  const state = getSession(gameId);
  if (!state) {
    res.status(404).json({ error: "Game not found." });
    return;
  }
  if (state.creatorId !== body.playerId) {
    res.status(403).json({ error: "Only the creator can start the game." });
    return;
  }
  if (state.status !== "pending") {
    res.status(400).json({ error: "Game has already started." });
    return;
  }

  // Generate resources for all countries via AI
  const resources = await generateCountryResources(
    Object.fromEntries(Object.keys(state.countries).map((name) => [name, {}])),
  );

  Object.entries(resources).map(([countryName, res]) => {
    if (state.countries[countryName]) {
      state.countries[countryName].favoriteResource = res.favoriteResource;
      state.countries[countryName].hatedResource = res.hatedResource;
    }
  });

  state.status = "active";
  updateState(gameId, state);
  res.json({ status: "started" });
};

lobbyRoutes.post("/game/create", createGame);
lobbyRoutes.post("/game/:gameId/join", joinGame);
lobbyRoutes.post("/game/:gameId/start", startGame);
